# 01 Importing Libraries

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

# 02 Importing Data


In [4]:
# importing Shortcut
path = r'C:\Users\georg\Documents\2025-02-05 Instacart Basket Analysis'

In [7]:
# importing Product DF
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [8]:
# importing Orders DF
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03 Creating a Test Dataframe

In [11]:
df_test = pd.DataFrame()

In [13]:
#Create a mixed type col
df_test['mix'] = ['a', 'b', 1, True]

In [16]:
# check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [17]:
df_test['mix'] = df_test['mix'].astype('str')

In [21]:
df_test.dtypes

mix    object
dtype: object

# 04 Missing Values

In [22]:
# finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [23]:
# looking into missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [24]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [26]:
# replace missing values with mean: df['column with missings'].fillna(mean value, inplace=True)
# replace missing values with median: df['column with missings'].fillna(median value, inplace=True)

In [28]:
df_prods.shape

(49693, 5)

In [29]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [30]:
df_prods_clean.shape

(49677, 5)

In [31]:
# anotehr code to drop missing values: df_prods.dropna(inplace = True)
# if we wanted to drop NaNs from a particular col: df_prods.dropna(subset = ['product_name'], inplace = True)

# 05 Duplicates

In [33]:
# create a subset of df prods clean - containing only rows that are duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [34]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [35]:
# before deleting, check the number of rows
df_prods_clean.shape

(49677, 5)

In [36]:
# delete duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [37]:
df_prods_clean_no_dups.shape

(49672, 5)

### 5 less rows, which makes sense since table on top shows 5 duplicates

# 06 Exercise

In [39]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  time_of_order  days_since_prior_order  
count       3.421083e+06   3.421083e+06            3.214874e+06  
mean        2.776219e+00   1.345202e+01            1.111484e+01  
std         2.046829e+00   4.226088e+00            9.206737e+00  
min         0.000000e+00   0.000000e+00            0.000000e+00  
25%         1.000000e+00   1.000000e+01            4.000000e+00  
50%         3.000000e+00   1.300000e+01            7.000000e+00  
75%         5.000000e+00   1.600000e+01            1.500000e+01  
max         6.000000e+00   2.300000e+01            3.000000e+01

### Nothing looks out of the ordinary.

In [41]:
# check for mixed types
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

### no mixed-type columns in the Order DF

In [42]:
# finding missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
time_of_order                  0
days_since_prior_order    206209
dtype: int64

### the days since prior order column have 206,209 missing values. This might be because of first time orders. I would need to look into it to find out whats actually going on

In [43]:
# looking into missing values
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [44]:
df_nan_ords

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         time_of_order  days_since_prior_order  
0                    8                     NaN  
11                  11                     NaN  
26                  14                     NaN  
39                  11                     NaN  
45                  12                     NaN  
...                ...                     ...  
3420930             12                     NaN  
3420934             18                     NaN  
3421002             19                     NaN  
3421019             15                     NaN  
3421069             11                     NaN  

[206209 rows x 7 columns]

### this looks like it might be each customers first order

In [49]:
#highlighting the user id col in the above table
df_nan_ords_user_ids = df_nan_ords.get('user_id')
df_nan_ords_user_ids

0               1
11              2
26              3
39              4
45              5
            ...  
3420930    206205
3420934    206206
3421002    206207
3421019    206208
3421069    206209
Name: user_id, Length: 206209, dtype: int64

In [50]:
#looking for duplicates in user id col
df_nan_ords_dups = df_nan_ords_user_ids[df_nan_ords_user_ids.duplicated()]

### no duplicates in the user id col. So that means that the missing values are due to them being first time orders

In [51]:
#looking for duplicates DF ords
df_ords_dups = df_ords[df_ords.duplicated()]

### no duplicates in DF ords

In [54]:
# Exporting new ords DF
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [55]:
# Exporting new prods DF
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))